src/core/textrenderer/qgsfontmanager.cpp

```diff
@@ -0,0 +1,1122 @@
+    if ( id == -1 )
+    {
+      errorMessage = tr( "Could not install font from %1" ).arg( destPath );
+      return false;
+    }
+    else
+    {
+      locker.changeMode( QgsReadWriteLocker::Write );
+      mUserFontToFamilyMap.insert( destPath, foundFamilies );
+      mUserFontToIdMap.insert( destPath, id );
+    }
+    return true;
+  }
+  else
+  {
+    // font install failed, but maybe it's a zip file
+    QStringList files;
+    if ( QgsZipUtils::unzip( tempFile.fileName(), tempDir.path(), files ) )
+    {
+      locker.changeMode( QgsReadWriteLocker::Write );
+      for ( const QString &file : std::as_const( files ) )
+      {
+        const QFileInfo fi( file );
+        if ( fi.fileName().compare( QLatin1String( "OFL.txt" ), Qt::CaseInsensitive ) == 0
+             || fi.fileName().compare( QLatin1String( "LICENSE.txt" ), Qt::CaseInsensitive ) == 0 )
+        {
+          QFile licenseFile( file );
+          if ( licenseFile.open( QIODevice::ReadOnly ) )
+          {
+            QTextStream in( &licenseFile );
+            const QString license = in.readAll();
+            licenseDetails.append( license );
+          }
+        }
+        else if ( fi.suffix().compare( QLatin1String( "ttf" ), Qt::CaseInsensitive ) == 0 ||
+                  fi.suffix().compare( QLatin1String( "otf" ), Qt::CaseInsensitive ) == 0 )
+        {
+          sourcePath = file;
+          id = QFontDatabase::addApplicationFont( sourcePath );
+          if ( id != -1 )
+          {
+            QFontDatabase::removeApplicationFont( id );
+            const QString destPath = fontsDir.filePath( fi.fileName() );
+            if ( !QFile::copy( sourcePath, destPath ) )
+            {
+              errorMessage = tr( "Could not copy font to %1" ).arg( destPath );
+              return false;
+            }
+            id = QFontDatabase::addApplicationFont( destPath );
+            if ( id == -1 )
+            {
+              errorMessage = tr( "Could not install font from %1" ).arg( destPath );
+              return false;
+            }
+            const QStringList foundFamilies = QFontDatabase::applicationFontFamilies( id );
+            mUserFontToFamilyMap.insert( destPath, foundFamilies );
+            mUserFontToIdMap.insert( destPath, id );
+            for ( const QString &found : foundFamilies )
+            {
+              if ( !families.contains( found ) )
+                families << found;
+            }
+          }
+        }
+      }
+      return true;
+    }
+  }
+
+  errorMessage = tr( "Could not read fonts from data" );
+  return false;
+}
+
+void QgsFontManager::addUserFontDirectory( const QString &directory )
+{
+  QgsReadWriteLocker locker( mReplacementLock, QgsReadWriteLocker::Read );
+  if ( mUserFontDirectories.contains( directory ) )
+    return;
+
+  locker.changeMode( QgsReadWriteLocker::Write );
+  mUserFontDirectories.append( directory );
+  locker.unlock();
+  installUserFonts();
```

Comment:
Probably overthinking it here but I'm wondering whether we should avoid re-scanning/installing _all_ user fonts of all registered directories when adding a new one. We could have a private function to adds fonts from a given directory parameter, rely on that in the installUserFonts() function, and call the private function here instead? 

I mean, I doubt people will have font directories with thousands of fonts, so this could all well not be worth the effort :)

Reply:
Fixed